File: _programmers/_14_/network.js
// https://programmers.co.kr/learn/courses/30/lessons/43162

const dfs = (n, computers, visited, start) => {
  const stack = [start]
  let i
  while (stack.length) {
    i = stack.pop()

    if (visited[i] == 0) {
      visited[i] = 1
    }
    [...Array(n).keys()].forEach(j => {
      if ((computers[i][j] == 1) && (visited[j] == 0)) {
        stack.push(j)
      }
    })
  }
}

function solution(n, computers) {
  const visited = Array(n).fill(0);
  let answer = 0;

  [...Array(n).keys()].forEach(i => {
    if (! visited[i]) {
      dfs(n, computers, visited, i)
      answer += 1
    }
  })
  return answer
}

console.log(solution(3, [
  [
    1, 1, 0
  ],
  [
    1, 1, 1
  ],
  [
    0, 1, 1
  ]
]))

File: _programmers/_14_/network.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

def dfs(n, computers, visited, start):
  stack = [start]
  while stack:
    j = stack.pop()
    if visited[j] == 0:
      visited[j] = 1
    for i in range(n):
      if computers[j][i] == 1 and visited[i] == 0:
        stack.append(i)
  return

def solution(n, computers):
  answer = 0
  visited = [0] * n

  for i in range(n):
    if not visited[i]:
      dfs(n, computers, visited, i)
      answer += 1
  return answer

print(
  solution(
    3, 
    [
      [1, 1, 0], 
      [1, 1, 0], 
      [0, 0, 1]
    ]	
  )
)